### PR TITLE
SLING-10150 improve handling of multivalue hides

### PR DESCRIPTION
**Pull request by kwin**
affects both hideChildren and hideProperties

### CODE REVIEW

File: src/main/java/org/apache/sling/resourcemerger/impl/MergingResourceProvider.java [hunk 1]
```diff
@@ -51,86 +51,55 @@
...
          */
-        public ParentHidingHandler(final Resource parent, final boolean traverseParent) {
+        public ResourceHidingHandler(final Resource resource, final boolean traverseParent) {
+            isParentHiddenFully = false;
+            isParentHiddenForUnderlay = false;
             // evaluate the sling:hideChildren property on the current resource
-            final ValueMap parentProps = parent.getValueMap();
-            final String[] childrenToHideArray = parentProps.get(MergedResourceConstants.PN_HIDE_CHILDREN, String[].class);
+            final ValueMap properties = resource.getValueMap();
+            final String[] childrenToHideArray = properties.get(MergedResourceConstants.PN_HIDE_CHILDREN, String[].class);
```
Comment (jsedding) [inline src/main/java/org/apache/sling/resourcemerger/impl/MergingResourceProvider.java:74]: If you call `properties.get(MergedResourceConstants.PN_HIDE_CHILDREN, new String[0])` instead, you can eliminate the following null check and the one for `hidePredicate != null` further down.

**Review by jsedding — APPROVED** [review]
Looks good if I understood everything correctly. Just nitpicks, except for the escaping issue I mention. But that could also be tackled in a separate issue if you prefer.

**Review by henrykuijpers — CHANGES_REQUESTED** [review]
Please integrate the added unit test of #3 as well.

**Comment by kwin** [comment]
> Please integrate the added unit test of #3 as well.

Done in https://github.com/apache/sling-org-apache-sling-resourcemerger/pull/4/commits/ab7a679579dd231d4f81ca2ed6d5d2acd6d99b5f

File: pom.xml [hunk 2]
```diff
@@ -130,5 +134,10 @@
             <version>4.1</version>
             <scope>test</scope>
         </dependency>
+        <dependency>
+            <groupId>org.slf4j</groupId>
+            <artifactId>slf4j-simple</artifactId>
+             <scope>test</scope>
```
[the remark this answers is not in the document]
Comment (kwin) [reply pom.xml]: Fixed in https://github.com/apache/sling-org-apache-sling-resourcemerger/pull/4/commits/32cf5745f8dce1246209396d631a96445d49adbb

File: src/main/java/org/apache/sling/resourcemerger/impl/HideItemPredicate.java [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+    private static final Logger LOGGER = LoggerFactory.getLogger(HideItemPredicate.class);
+
+    public HideItemPredicate(String[] settings, String propertyPath) {
+        names = new HashSet<>();
+        // negated and non-negated values must not be mixed
+        for (String setting : settings) {
+            if (setting.equals("*")) {
+                isWildcard = true;
+            } else {
+                boolean isNegated = setting.startsWith("!") && !setting.startsWith("!!");
```
[the remark this answers is not in the document]
Comment (kwin) [reply src/main/java/org/apache/sling/resourcemerger/impl/HideItemPredicate.java]: Fixed in https://github.com/apache/sling-org-apache-sling-resourcemerger/pull/4/commits/32cf5745f8dce1246209396d631a96445d49adbb

File: src/main/java/org/apache/sling/resourcemerger/impl/MergingResourceProvider.java (hunk 1 above)
Comment (kwin) [reply src/main/java/org/apache/sling/resourcemerger/impl/MergingResourceProvider.java]: Fixed in https://github.com/apache/sling-org-apache-sling-resourcemerger/pull/4/commits/32cf5745f8dce1246209396d631a96445d49adbb

**Pull request merged**